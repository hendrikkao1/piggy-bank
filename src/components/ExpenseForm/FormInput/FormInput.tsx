import React from "react";
import { useFormContext } from "react-hook-form";
import { Input } from "@/components/Input/Input";
import { IExpenseFormData } from "../ExpenseForm.types";
import { FromFieldError } from "../FromFieldError/FromFieldError";

interface FormInputProps extends React.ComponentPropsWithoutRef<"input"> {
  field: keyof IExpenseFormData;
}

export const FormInput = ({ field, ...rest }: FormInputProps) => {
  const { register, formState } = useFormContext<IExpenseFormData>();

  const error = formState.errors[field];

  return (
    <FromFieldError error={error?.message}>
      <Input aria-invalid={Boolean(error)} {...register(field)} {...rest} />
    </FromFieldError>
  );
};
