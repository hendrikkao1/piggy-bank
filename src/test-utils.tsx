import React from "react";
import {
  RenderHookOptions,
  RenderOptions,
  render,
  renderHook,
} from "@testing-library/react";
import { NextIntlClientProvider } from "next-intl";
import messages from "./messages/en-US.json";

export const AllTheProviders = ({
  children,
}: {
  children: React.ReactNode;
}) => (
  <NextIntlClientProvider locale={"en-US"} messages={messages}>
    {children}
  </NextIntlClientProvider>
);

const customRender = (ui: React.ReactElement, options: RenderOptions = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = <TProps, TResult>(
  callback: (props: TProps) => TResult,
  options?: RenderHookOptions<TProps>,
) => renderHook(callback, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";

export { customRender as render, customRenderHook as renderHook };
