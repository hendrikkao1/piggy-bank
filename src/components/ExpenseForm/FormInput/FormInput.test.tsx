import { render } from "@/test-utils";
import { FormInput } from "./FormInput";
import { UseFormReturn } from "react-hook-form";
import { useFormContext } from "react-hook-form";

jest.mock("react-hook-form", () => ({
  ...jest.requireActual("react-hook-form"),
  useFormContext: jest.fn(),
}));

const mockUseFormContext = useFormContext as jest.MockedFunction<
  typeof useFormContext
>;

describe("FormInput", () => {
  describe("aria-invalid", () => {
    it("have `false` when there is error", () => {
      mockUseFormContext.mockReturnValue({
        register: jest.fn(),
        formState: {
          errors: {},
        },
      } as unknown as UseFormReturn);

      const { container } = render(<FormInput field="amount" />);

      expect(container).toMatchSnapshot();
    });

    it("have `true` when there is an error", () => {
      mockUseFormContext.mockReturnValue({
        register: jest.fn(),
        formState: {
          errors: {
            amount: {
              message: "amount.required",
            },
          },
        },
      } as unknown as UseFormReturn);

      const { container } = render(<FormInput field="amount" />);

      expect(container).toMatchSnapshot();
    });
  });
});
