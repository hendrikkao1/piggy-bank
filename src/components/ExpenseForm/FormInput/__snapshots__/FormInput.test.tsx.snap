// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput aria-invalid have \`false\` when there is error 1`] = `
<div>
  <input
    aria-invalid="false"
    class="bg-white border border-zinc-900/20 text-zinc-900 text-sm rounded-full block w-full px-4 py-2 aria-[invalid=true]:bg-red-100 aria-[invalid=true]:border-red-900  aria-[invalid=true]:text-red-900"
  />
</div>
`;

exports[`FormInput aria-invalid have \`true\` when there is an error 1`] = `
<div>
  <input
    aria-invalid="true"
    class="bg-white border border-zinc-900/20 text-zinc-900 text-sm rounded-full block w-full px-4 py-2 aria-[invalid=true]:bg-red-100 aria-[invalid=true]:border-red-900  aria-[invalid=true]:text-red-900"
  />
  <div
    class="mt-2"
  >
    <p
      class="text-xs text-red-900"
      role="alert"
    >
      Amount is a mandatory field, for example 123.45
    </p>
  </div>
</div>
`;
